### PR TITLE
Putting the quick in quick tree: compile an instantiation function for each class model type

### PR DESCRIPTION
**Pull request by airhorns**
I was curious about how hard and how effective the "compile a function that does the same thing but with loops unrolled and indirection removed" approach would be for model instantiation. I tried it. It works ... really well.

On `main`:

```
❯ p x bench/create-large-root.ts

> @gadgetinc/mobx-quick-tree@0.4.0 x /Users/airhorns/Code/mobx-quick-tree
> ts-node --transpile-only "bench/create-large-root.ts"

instantiating a large root x 395 ops/sec ±0.88% (90 runs sampled)
```

On the first pass optimization branch in #44:

```
❯ p x bench/create-large-root.ts

> @gadgetinc/mobx-quick-tree@0.4.0 x /Users/airhorns/Code/mobx-quick-tree
> ts-node --transpile-only "bench/create-large-root.ts"

instantiating a large root x 925 ops/sec ±1.01% (88 runs sampled)
```

On this branch:

```
❯ p x bench/create-large-root.ts

> @gadgetinc/mobx-quick-tree@0.4.0 x /Users/airhorns/Code/mobx-quick-tree
> ts-node --transpile-only "bench/create-large-root.ts"

instantiating a large root x 3,182 ops/sec ±0.26% (97 runs sampled)
```

I am surprised and delighted 💨 💨 💨 . I can also confirm this is real, I pushed a prerelease to Gadget to see if I was smoking performance-enhancing drugs and https://github.com/gadget-inc/gadget/pull/6444 is green!

The old approach that instantiated every class model looked like this: 

```typescript
export const instantiateInstanceFromProperties = (
  instance: any,
  snapshot: Record<string, any> | undefined,
  properties: ModelProperties,
  identifierProp: string | undefined,
  context: InstantiateContext
) => {
  for (const propName in properties) {
    const propType = properties[propName];
    if (isReferenceType(propType.mstType)) {
      context.referencesToResolve.push(() => {
        const propValue = propType.instantiate(snapshot?.[propName], context);
        instance[propName] = propValue;
      });
      continue;
    }

    const propValue = propType.instantiate(snapshot?.[propName], context);
    setParent(propValue, instance);
    instance[propName] = propValue;
  }

  if (identifierProp) {
    const id = instance[identifierProp];
    Object.defineProperty(instance, $identifier, { value: id });
    context.referenceCache.set(id, instance);
  }
};
```

which is to say, loop over the `SomeClassModel.properties` object which stores `{ key: SimpleType, createdDate: DateType, childObjects: MapType}` etc, grabbing each key/value of the snapshot and using the type from the `.properties` to go from snapshot => live MQT object. This snapshot => live MQT object transform is what the `.instantiate` function on all our types is responsible for. Some examples:
 - on strings and numbers, do nothing
 - on dates, convert the incoming snapshot epoch number to a date with `new Date()`
 - on optional types, check for undefined / null and use the default if the snapshot is that, otherwise, instantiate the actual type 
 - on map types, create an instance for each of the snapshot values, then assemble them into a `QuickMap` map 

So, our compiler emits inline JS to do each of those things for the types we've fast-pathed, and then falls back on the existing `.instantiate` implementations for anything we haven't.

Here's what a compiled instantiation function looks like:

```javascript
function buildCustomerInstantiator(model, imports) {
  const { QuickMap, $identifier } = imports;
  const v0 = model.properties["email"];
  const v1 = model.properties["phoneNumber"];

  return function InstantiateCustomer(instance, snapshot, context) {
    // optional type for type
    let snapshotValuetype = snapshot?.["type"];
    if (snapshotValuetype === undefined) {
      snapshotValuetype = "Customer";
    }

    instance["type"] = snapshotValuetype;

    // simple type for key
    instance["key"] = snapshot?.["key"];

    // simple type for firstName
    instance["firstName"] = snapshot?.["firstName"];

    // simple type for lastName
    instance["lastName"] = snapshot?.["lastName"];

    // instantiate fallback for email of type (string | null)
    instance["email"] = v0.instantiate(snapshot?.["email"], context, instance);

    // instantiate fallback for phoneNumber of type (string | null)
    instance["phoneNumber"] = v1.instantiate(
      snapshot?.["phoneNumber"],
      context,
      instance
    );

    // optional type for createdAt
    let snapshotValuecreatedAt = snapshot?.["createdAt"];
    if (snapshotValuecreatedAt === undefined) {
      snapshotValuecreatedAt =
        model.properties["createdAt"].defaultValueOrFunc();
    }

    instance["createdAt"] = snapshotValuecreatedAt;

    const id = instance["key"];
    instance[$identifier] = id;
    context.referenceCache.set(id, instance);
  };
}
```

The weird outer function is an artifact of the way that you are supposed to do `eval` in JS these days. `eval` is the sketchiest and slowest of the options -- better is to use `new Function(<source>)`, which creates a new function in the global scope instead of the local one. This is good because you can't accidentally close over a bunch of stuff and leak it, but bad because if you do need context in the created function, you have to inject it weirdly. So, we actually create two functions: an outer one which establishes any variables we need in scope, like types or the `QuickMap` constructor, and then an inner function which is the one we actually run over and over and is the thing that needs to be very fast. Not super important but does make things look a bit weird.

There's a couple downsides of this whole approach:
 - the fast-path implementations could diverge from the slow-path, leading to hard to diagnose bugs
 - different type arrangements might expose end-user types to different paths over time
 - the fast-paths do a bit less less data validation, and may allow incorrect snapshots silently. We actually already accepted this for MQT, and rely on MST to do typechecking at write time, so MQT already doesn't do much runtime type validation at all, but the fast paths do even less.
 
Personally, I've found that the instantiate implementations haven't changed that much -- we've changed how we structure it all, but we haven't changed the core ideas at all. MQT since inception has had the same idea of an optional type, a map, of an array, of a model, etc, and I don't expect much churn in the list of primitives. I think we can expect some churn in how we implement them to be incredibly fast, but because the requirements are relatively set, I am ok adding brittleness to get that speed.

### CODE REVIEW

File: src/fast-instantiator.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    const aliasFuncBody = `
+    const { QuickMap, $identifier } = imports;
+    ${Array.from(this.aliases.entries())
+      .map(([expression, alias]) => `const ${alias} = ${expression};`)
+      .join("\n")}
+
+    ${innerFunc}
+  `;
+
+    // console.log(`function for ${this.model.name}`, "\n\n\n", aliasFuncBody, "\n\n\n");
```
Comment (thegedge) [inline src/fast-instantiator.ts:91]: 🔥 ?

File: src/fast-instantiator.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    if (identifierProp) {
+      segments.push(`
+      const id = instance["${identifierProp}"];
+      instance[$identifier] = id;
+      context.referenceCache.set(id, instance); 
+    `);
+    }
+
+    const innerFunc = `
+      return function Instantiate${this.model.name}(instance, snapshot, context) {
```
Comment (thegedge) [inline src/fast-instantiator.ts:77]: We'll get a nice function name in our stack traces, eh? 🚀 

Should we `camelize(name)` or something though? I can't remember if MST had any strict requirements on the name for a `types.model`. Or will this be the name of a class, which already adheres to the JS function/var/etc naming constraints?

**Review by thegedge — APPROVED** [review]
I'm quite surprised at the extent of the improvement by "inlining" the instantiation, but super game!

I do know, from recently doing some of the advent of code problems, that `for(const of abc of xyz)` is actually quite slow compared to `for(let index …) { const abc = xyz[index] }`, so maybe it shouldn't be surprising 😅

File: src/fast-instantiator.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+/**
+ * Compiles a fast function for taking snapshots and turning them into instances of a class model.
+ **/
+export const buildFastInstantiator = <T extends IClassModelType<Record<string, IAnyType>, any, any>>(model: T): CompiledInstantiator<T> => {
+  return new InstantiatorBuilder(model).build();
+};
+
+type DirectlyAssignableType = SimpleType<any> | IntegerType | LiteralType<any> | DateType;
+const isDirectlyAssignableType = (type: IAnyType): type is DirectlyAssignableType =>
+  type instanceof SimpleType || type instanceof IntegerType || type instanceof LiteralType || type instanceof DateType;
```
[the remark this answers is not in the document]
Comment (airhorns) [reply src/fast-instantiator.ts:27]: I don't think that the case -- maybe it was, but all three extend BaseType right now!

File: src/fast-instantiator.ts (hunk 2 above)
Comment (airhorns) [reply src/fast-instantiator.ts:77]: Right now we're only using this for class models which ya will have a safe JS function name so I think this is ok. We also don't actually have a camelize utility handy, just the one memoize function from lodash atm.

File: src/fast-instantiator.ts (hunk 1 above)
Comment (airhorns) [reply src/fast-instantiator.ts:91]: I found this was really handy for debugging the generated code so I figured I'd leave it for the next person?

File: src/fast-instantiator.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      context.referencesToResolve.push(() => {
+        if (${varName}) {
+          const referencedInstance = context.referenceCache.get(${varName});
+          if (referencedInstance) {
+            instance["${key}"] = referencedInstance;
+            return;
+          }
+        }
+        ${notFoundBehavior}
+      });
```
[the remark this answers is not in the document]
Comment (airhorns) [reply src/fast-instantiator.ts:123]: I think this is a good idea, but I'd rather wait for some churn that suggests what exactly should be shared and what shouldn't right now. There's a few optimizations that the codegen can make by unrolling all the functions into one and by referencing stuff from the outer context that I think would make it hard to pull off the sharing thing, but I do worry that the implementations might diverge. I think the Gadget codebase serves as a really good integration test of the whole thing so I feel comfortable with it being a side-by-side implementation of the instantiation logic for now.

File: src/fast-instantiator.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    if (type.defaultValueOrFunc instanceof Function) {
+      defaultValueExpression = `model.properties["${key}"].defaultValueOrFunc()`;
+    } else {
+      defaultValueExpression = JSON.stringify(type.defaultValueOrFunc);
+    }
+
+    const varName = `snapshotValue${key}`;
+
+    const comparisonsToUndefinedValues = (type.undefinedValues ?? [undefined]).map((value) => {
+      return `(${varName} === ${JSON.stringify(value)})`;
```
[the remark this answers is not in the document]
Comment (airhorns) [reply src/fast-instantiator.ts:146]: Good call, fixing!

File: src/fast-instantiator.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (thegedge) [reply src/fast-instantiator.ts:27]: Ah, perhaps I was accidentally looking at an old version

**Comment by airhorns** [comment]
ping @angelini

**Review by angelini — APPROVED** [review]
This is so cool 👏 

I don't have a ton of familiarity with MQT internals, but this is really readable considering what it's doing.

File: src/fast-instantiator.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (thegedge) [reply src/fast-instantiator.ts:123]: Yeah, I think we can cross that bridge when we get to it. Specifically, if we have a divergence that we miss, we can come back and think about this 👍

**Pull request merged**